src/inmanta/server/services/compilerservice.py

```diff
@@ -417,14 +417,22 @@ def _compile_merge_key(c: data.Compile) -> Hashable:
 
     async def _queue(self, compile: data.Compile) -> None:
         async with self._global_lock:
+            env: Optional[data.Environment] = await data.Environment.get_by_id(compile.environment)
+            assert env is not None
```

Comment:
A long time ago, we had a discussion about when assertions should be used and when not. If I remember everything well, the conclusion was that assertions should be used to give the reader of the code access to certain invariant that should hold at all times according to the developer. This implies that assertions should never evaluate to False in production. That's important because assertions can be turned off in the python interpreter. This usage violates that rule, because an environment might be deleted by the time this code executes. @wouterdb Wdyt?

Reply:
That is correct. 

It would be better to write this in a form that can not be disabled. (if env is None: raise AssertionError())

Excellent catch.

Reply:
Right. I remembered the discussion but I hadn't thought about the possibility of the environment being deleted in between. I'll raise an explicit exception.